deduce.go

```diff
@@ -690,6 +695,253 @@ func (sm *SourceMgr) deduceFromPath(path string) (deductionFuture, error) {
 	}, nil
 }
 
+type deductionCoordinator struct {
+	ctx      context.Context
```

Comment:
@rogpeppe has also [responded](https://www.reddit.com/r/golang/comments/60gm4g/constext_cons_your_contexts/df7y8dd/) on reddit, with a hopefully-clearer explanation :)

> Are you concerned that there are multiple entrypoints to this code, and you want to be sure the caller is using the same context for them? If so, I think the right approach is to shrink the entrypoints.

Yes. I don't see how it would be possible to shrink the entrypoints - there's the creation of the `sourceManager`, and there's calling methods on it - the context for the former cancels ALL running activities, and the context for the latter cancels only the current method. 

These contexts are probably from the same tree, but we cannot know if they are from within the `sourceMgr`, so we have to take precautions.

> There's no need for quitchan there - you'd just call `cancel()` at the end of the function directly.

Without the quitchan, we don't know to cancel this running call if the long-lived context was canceled. We'd have to make the assumption that the two contexts are part of the same tree.

> If you want to only exit once all spawned goroutines have exited, you keep track of that with `sync.WaitGroup`.

Yep, but this is entirely separate.

> I just think it's simpler to think about the one flow of contexts than this idea of merging two.

I agree, if we could make that assumption, it would be the simpler way to do it :)